Added support for date insertion in localized files

tombogle opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/SIL.BuildTasks/76)
<!-- Reviewable:end -->

tombogle committed:
Added support for date insertion in localized files

ermshiperete commented on SIL.BuildTasks.Tests/FileUpdateTests.cs:
```diff
@@ -62,6 +63,120 @@ public void GetModifiedContents_RegexTextNotMatched_Throws(string origContents,
 			Assert.That(ex.Message, Is.EqualTo($"No replacements made. Regex: '{regex}'; ReplacementText: '{replacement}'"));
 		}
 
+		[TestCase("_DATE_ _VERSION_\r\nStuff", "_DATE_", "M/yyyy", "{0} 3.2.1\r\nStuff")]
```
This would be nicer if we'd use the [ReproducibleDateTimeProvider](https://github.com/sillsdev/libpalaso/blob/master/SIL.TestUtilities/Providers/ReproducibleDateTimeProvider.cs). Then we could do:

```suggestion
		[TestCase("_DATE_ _VERSION_\r\nStuff", "_DATE_", "M/yyyy", "January/2025 3.2.1\r\nStuff")]
```

But that would introduce a new dependency...

tombogle replied:
Worth considering.

tombogle replied:
I went ahead and did it. It does simplify the tests. Now that it's done, you can re-review and consider whether anything bad is likely to come of the added dependency. I think it should be okay.
<!-- Reviewable comment -ORHWTYJ2gYdPwhHgeIG:b-w7uci -->
<!-- Sent from Reviewable.io -->

ermshiperete commented on SIL.BuildTasks.Tests/FileUpdateTests.cs:
```diff
@@ -62,6 +63,120 @@ public void GetModifiedContents_RegexTextNotMatched_Throws(string origContents,
+			};
+
+			var currentDate = string.Format(DateTime.UtcNow.Date.ToString("d {0} yyyy"),
+				GetMonthName(locale, DateTime.UtcNow.Month));
```
This would be slightly easier:

```suggestion
			var currentDate = DateTime.UtcNow.Date.ToString("d MMMM yyyy", new CultureInfo(locale));
```

tombogle replied:
True. I was intentionally trying to make the test be as much as possible NOT a replication of the production code and trying to illustrate clearly in the test what the point of the locale logic is. I was considering when I wrote GetMonthName the possibility of implementing it by hard-coding the month names.

tombogle replied:
With the use of the ReproducibleDateTimeProvider, I was able to just hardcode the expected month names.
<!-- Reviewable comment -ORHWmQSDGcTToJAc_m2:b9ypcrd -->
<!-- Sent from Reviewable.io -->

tombogle committed:
Merge branch 'master' into allow-in-file-date-format-specifiers

tombogle committed:
Used DateTimeProvider and (in tests) ReproducibleDateTimeProvider to simplify unit tests

Note that I intentionally used a very old version of the new dependencies because it was all I needed and I figured it might help avoid conflicts with consumers of this package that wanted to stay back.

tombogle reviewed (commented):
<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
<!-- Sent from Reviewable.io -->

tombogle committed:
Replaced SuppressMessage attributes with PublicAPI from JetBrains.Annotations

Deprecated AbandondedSuites in favor of correctly spelled AbandonedSuites
Misc minor code refactoring and cleanup.
Fixed typos in comments.

tombogle committed:
Fixed copyright date ranges in source files and added note about case-change in CHANGELOG

ermshiperete commented on SIL.BuildTasks/SIL.BuildTasks.csproj:
```diff
@@ -9,6 +9,7 @@
   <ItemGroup>
     <PackageReference Include="Markdig.Signed" Version="0.41.1" />
     <PackageReference Include="Microsoft.Build.Tasks.Core" Version="17.14.8" />
+    <PackageReference Include="SIL.Core" Version="7.0.0" />
```
Why version 7.0 and not the latest 16.0?

tombogle replied:
As I wrote in the commit comment: "Note that I intentionally used a very old version of the new dependencies because it was all I needed and I figured it might help avoid" If you think this is not a desirable thing to do, I can upgrade, but that was my rationale.
<!-- Reviewable comment -ORM8Auj5wJQ0HMwE1t1:bgtrz0 -->
<!-- Sent from Reviewable.io -->

tombogle replied:
...conflicts with consumers of this package that wanted to stay back.
<!-- Reviewable comment -OXTf0x3BZteCEbuW2aY:bhgjhj9 -->
<!-- Sent from Reviewable.io -->

tombogle reviewed (commented):
<!-- Sent from Reviewable.io -->

tombogle committed:
Merge branch 'master' into allow-in-file-date-format-specifiers

tombogle reviewed (commented):
<!-- Sent from Reviewable.io -->

tombogle commented:
@ermshiperete If you want to give this a final review, that would be great.
